acq_stat_reports/utils.py

```diff
@@ -0,0 +1,538 @@
+
+    x = _mpl_hist_steps(
+        quantiles[f"{col}_low"],
+        quantiles[f"{col}_high"],
+    )
+    sigma_1_low = _mpl_hist_steps(sigma_1_low_frac)
+    sigma_2_low = _mpl_hist_steps(sigma_2_low_frac)
+    sigma_1_high = _mpl_hist_steps(sigma_1_high_frac)
+    sigma_2_high = _mpl_hist_steps(sigma_2_high_frac)
+
+    sigma_band_1 = plt.fill_between(
+        x, sigma_1_low, sigma_1_high, color="gray", alpha=0.8
+    )
+    sigma_band_2 = plt.fill_between(
+        x, sigma_2_low, sigma_2_high, color="gray", alpha=0.3
+    )
+
+    bins = bins[np.isfinite(bins)]
+    plt.xlim(bins[0], bins[-1])
+    plt.ylim(-0.05, 1.05)
+
+    labels = {
+        scatter: "Observed",
+        sigma_band_1: "68.2% range",
+        sigma_band_2: "95.4% range",
+    }
+    plt.legend(
+        labels.keys(),
+        labels.values(),
+        loc="best",
+    )
+
+
+@mpl_plot(
+    title="Acquisition Probability",
+    xlabel="Expected p$_{acq}$",
+    ylabel="Observed p$_{acq}$",
+    figscale=(1, 2),
+)
+def binned_data_probability_plot(
+    binned_data: BinnedData,
+    **kwargs,  # noqa: ARG001 (kwargs is neeeded by the decorator)
+):
+    quantiles = binned_data.binned_data
+
+    sel = quantiles["n"] > 0
+    prob_exp = quantiles["p_acq_model_mean"][sel]
+    prob_obs, low, high = binomial_confidence_interval(
+        quantiles["acqid"][sel], quantiles["n"][sel]
+    )
+
+    plt.gca().set_aspect("equal")
+    plt.errorbar(
+        prob_exp,
+        prob_obs,
+        yerr=[prob_obs - low, high - prob_obs],
+        xerr=quantiles["p_acq_model_std"][sel],
+        fmt=".",
+        color="k",
+        linewidth=1,
+    )
+    x = np.linspace(0, 1, 10)
+    plt.plot(x, x, "k--")
+    plt.ylim(0, 1)
+    plt.xlim(0, 1)
+
+
+def _get_quantiles_(p_acq_model, quantiles=None, n_realizations=10000):
+    """
+    Generate the 5th, 50th, and 95th percentiles of the expected number of successes in a series.
+
+    This function assumes the input is an array of probabilities, one for each draw.
+    It generates `n_realizations` realizations , and uses them to estimate the percentiles.
+
+    Parameters
+    ----------
+    p_acq_model : array-like
+        The binomial probability.
+    n_realizations : int, optional
+        The number of realizations of the series to use for the Monte Carlo estimation.
+
+    Returns
+    -------
+    array-like
+        The quantiles
+    """
+    if quantiles is None:
+        quantiles = [15.9, 50, 84.1]
+    samples = (
+        np.random.uniform(size=n_realizations * len(p_acq_model)).reshape(
+            (-1, len(p_acq_model))
+        )
+        < p_acq_model[None]
+    )
+    n = np.sum(samples, axis=1)
+    return np.percentile(n, quantiles)
+
+
+def get_histogram_quantile_ranges(  # noqa: PLR0915
+    data,
+    bin_edges,
+    extra_cols=(),
+    n_samples=10000,
+    success_column="acqid",
+    prob_column="p_acq_model",
+):
+    """
+    Summarize data in bins and calculate quantiles.
```

Comment:
This function handles exactly one dataset, so there might be a misunderstanding.

The looping over columns is to produce N-dimensional histograms. In other words, if you give bins with two columns, this function will not produce the (unrelated) columns corresponding to 1d histograms. It will produce a single 2d histogram.

Take a look at the test.